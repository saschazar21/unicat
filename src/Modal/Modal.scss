@import '../__styles__/variables';

@keyframes appear {
  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

@keyframes slideUp {
  0% {
    opacity: 0;
    transform: translateY(25vh);
  }

  100% {
    opacity: 1;
    transform: translateY(0);
  }
}

.backdrop,
.container {
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  overflow: hidden;
  z-index: 9998;
}

.backdrop {
  position: absolute;
  animation: appear $duration ease-out;
  background-color: rgba(color(default, false, $colors, true), 0.4);
}

.container {
  position: fixed;
  display: flex;
  justify-content: center;
  align-items: flex-end;

  @include media-query-up(m) {
    align-items: center;
  }
}

.wrapper {
  animation: slideUp $duration ease-out;
  display: block;
  border-bottom-right-radius: 0;
  border-bottom-left-radius: 0;
  box-sizing: border-box;
  margin: 0;
  max-height: calc(100vh - #{spacing(xl)});
  width: 100%;
  overflow-x: hidden;
  overflow-y: auto;
  z-index: 9999;

  @include media-query-up(m) {
    border-bottom-right-radius: borderRadius();
    border-bottom-left-radius: borderRadius();
    margin: spacing();
    max-width: rem(568px);
  }
}
