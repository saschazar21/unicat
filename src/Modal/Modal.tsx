import React, { Component, ReactNode, createRef, ReactElement } from 'react';
import classnames from 'classnames';

import { preventDefault } from '../__tools__/helpers';
import { SmallVariant } from '../__types__/global';
import Button from '../Button';
import Card from '../Card';

import styles from './Modal.scss';

export interface ModalProps {
  children: ReactNode[] | ReactNode;
  className?: string;
  cta?: ReactElement<Button>;
  heading?: ReactNode;
  onClose: (event?: React.SyntheticEvent) => void;
  variant?: SmallVariant;
}

export interface ModalState {
  overflow: string | null;
}

export default class Modal extends Component<ModalProps, ModalState> {
  static closingKeys = ['Backspace', 'Escape'];

  static defaultProps = {
    variant: 'default',
  };

  state = {
    overflow: '',
  };

  private containerRef: React.RefObject<HTMLDivElement> = createRef();

  private handleClick = (e: React.MouseEvent): void => {
    preventDefault(e);
    const { onClose } = this.props;

    onClose(e);
  };

  private handleKeyUp = (e: React.KeyboardEvent): void => {
    preventDefault(e);
    const { onClose } = this.props;

    if (e.key && Modal.closingKeys.indexOf(e.key) > -1) {
      onClose(e);
    }
  };

  public componentDidMount(): void {
    if (typeof document === 'object' && typeof document.body !== 'undefined') {
      this.setState({
        overflow: document.body.style.overflow,
      });
      document.body.style.overflow = 'hidden';

      if (this.containerRef && this.containerRef.current) {
        this.containerRef.current.focus();
      }
    }
  }

  public componentWillUnmount(): void {
    const { overflow } = this.state;

    if (typeof document === 'object' && typeof document.body !== 'undefined') {
      document.body.style.overflow = overflow;
    }
  }

  public render() {
    const { variant: defaultVariant } = Modal.defaultProps;

    const {
      children,
      className: customClassName,
      variant = defaultVariant,
      ...otherProps
    } = this.props;

    const className = classnames(styles.wrapper, customClassName);

    // Insert template logic here
    return (
      <div
        className={styles.container}
        onKeyUp={this.handleKeyUp}
        tabIndex={0}
        ref={this.containerRef}
      >
        <div className={styles.backdrop} onClick={this.handleClick} />
        <Card
          className={className}
          variant={variant as SmallVariant}
          {...otherProps}
        >
          {children}
        </Card>
      </div>
    );
  }
}
