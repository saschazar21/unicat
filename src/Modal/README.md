Renders a `<Card />` on a semi-black backdrop.

```jsx
import Button from '../Button';

initialState = { isOpen: false };
<div>
  <Button title="Open Modal" onClick={() => setState({ isOpen: true })} />
  {state.isOpen && <Modal onClose={() => setState({ isOpen: false })}><span>Hey I'm a Modal</span></Modal>}
</div>
```