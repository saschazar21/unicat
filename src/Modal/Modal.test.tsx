import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Modal from './Modal';


describe('Modal', () => {
  it('renders', () => {
    const f = jest.fn();
    const wrapper = shallow(
      <Modal onClose={f}>
        <span>Test</span>
      </Modal>
    );
    expect(toJson(wrapper)).toMatchSnapshot();

    expect(f).not.toHaveBeenCalled();

    const backdrop = wrapper.find('.backdrop');
    expect(backdrop).toHaveLength(1);

    backdrop.simulate('click');
    expect(f).toHaveBeenCalled();
  });

  it('closes using "esc" key', () => {
    const f = jest.fn();
    const wrapper = shallow(
      <Modal onClose={f}>
        <span>Test</span>
      </Modal>
    );

    expect(f).not.toHaveBeenCalled();
    wrapper.simulate('keyup', { key: 'Escape' });

    expect(f).toHaveBeenCalled();
  });
});