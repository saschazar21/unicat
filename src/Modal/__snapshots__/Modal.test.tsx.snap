// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders 1`] = `
<div
  className="container"
  onKeyUp={[Function]}
  tabIndex={0}
>
  <div
    className="backdrop"
    onClick={[Function]}
  />
  <Card
    className="wrapper"
    modal={true}
    onClose={[MockFunction]}
    variant="default"
  >
    <span>
      Test
    </span>
  </Card>
</div>
`;
